### PR TITLE
Update WMO BUFR format data and YAML files from single to multiple reports

### PR DESCRIPTION
**Pull request by gthompsnJCSDA**
## Description

The YAML and associated data files for processing 4 types of conventional observations from raw WMO BUFR format to IODA have been updated to go from single BUFR message examples to partial or full hour files that were captured using LDM (Local Data Manager).  This effort represents a straight passage from raw incoming data into IODA using Ron McLaren's bufr2ioda.x application, which can handle the way LDM adds carriage/line feed control characters and WMO headers that get ignored.

The input files were captured from an LDM server in real-time and represents how a stream of real-time software can operate from incoming data directly to IODA for the following data types:

- Surface land obs: SYNOP
- Surface marine obs:  Ship and buoy reports
- Aircraft obs: AMDAR

### Issue(s) addressed

A separate issue was not created.

## Acceptance Criteria (Definition of Done)

All ctests pass

## Dependencies

None

## Impact

No impacts to other repos.

### CODE REVIEW

**Review by rmclaren — CHANGES_REQUESTED** [review]
Code and YAML changes are fine, but we need to make sure that the unit tests continue to work (ie: we need to make sure the test output is up to date for the unit tests).

**Comment by gthompsnJCSDA** [comment]
> Code and YAML changes are fine, but we need to make sure that the unit tests continue to work (ie: we need to make sure the test output is up to date for the unit tests).

Yes, agreed. Like I mentioned in the FYI in description section.  I am circling back now to update prior files, which will take a bit.

**Comment by gthompsnJCSDA** [comment]
@srherbener   Can you suggest what can be done (simply) for many files in the testoutput directory that have the ```nlocs``` variable as a vector of all zeroes versus what happens as part of my PR that is creating an index from 1 to nlocs?

There are a number of files that my PR should not affect at all, so a whole bunch of ctests are failing for the above reason compared to those that I can fix due to my own code changes.  An example file is... ```iodaconv/test/testoutput/gds2_sst_l2p.nc``` which has all 63 nlocs vector as zeroes but the testrun file created on-the-fly has 1..63.

**Comment by rmclaren** [comment]
@gthompsnJCSDA I believe this is because your build of the "ioda" project is out of date. If you update this component and rebuild everything you should get a a vector of all 0's for nlocs and other dimensions.

**Comment by gthompsnJCSDA** [comment]
> @gthompsnJCSDA I believe this is because your build of the "ioda" project is out of date. If you update this component and rebuild everything you should get a a vector of all 0's for nlocs and other dimensions.


Makes sense, yes.  I think the built python libs are from a prior version so if I go up to the ioda-bundle level and rebuild, hopefully that does the trick.

**Comment by gthompsnJCSDA** [comment]
@rmclaren  Got all the testoutput files updated so now all tests are passing, yippee.  Note also my change to the datetime construct with seconds, because I was getting times with 1E+11 as the seconds due to missing seconds in some reports but not missing in others.

**Comment by rmclaren** [comment]
@gthompsnJCSDA I think @srherbener  could give a better answer, but I think since the missing value are marked with the fill value I think that this situation might be OK (having missing values marked as 10e10... So 0 those out might not really be necessary.

**Comment by gthompsnJCSDA** [comment]
> @gthompsnJCSDA I think @srherbener could give a better answer, but I think since the missing value are marked with the fill value I think that this situation might be OK (having missing values marked as 10e10... So 0 those out might not really be necessary.

Ron, I found it necessary because I have a full file of 1 hour of AMDAR aircraft data (globally) and some reports contain seconds while other reports do not.  So the string being made for the older datetime (string) variable was something like: 2021-08-01T00:21:1E+11Z which simply won't work.  So I forced in the zero values for anything outside [0,59].  I don't see a way around this for files that are a large concatenation of BUFR messages like in a real-time feed of data.

**Comment by rmclaren** [comment]
I see, your right. Forgot about that detail.

**Review by rmclaren — APPROVED** [review]

**Comment by rmclaren** [comment]
Looks good to me. I've approved it.

File: src/bufr/DataObject/ArrayDataObject.cpp
Comment (gthompsnJCSDA) [reply src/bufr/DataObject/ArrayDataObject.cpp:54]: Steve:  Ron and I had a discussion over email about this and there is a BUFR standard using just 1.E+11 as the indicator for missing value.  So that's why this works properly.  Otherwise, the ioda files were being given the 1.E+11 values as data while that line of code was saying the _FillValue was -999 so that was doing nothing useful downstream since it wasn't set to the proper missing value.  This change was something Ron addressed months ago in his feature branch that is awaiting other changes to submit.  Effectively he gave me a green light to pop in the fix and I adjusted all the test data outputs to match - so we're good to go.

File: src/bufr/BufrParser/Exports/Variables/DatetimeVariable.cpp
Comment (gthompsnJCSDA) [reply src/bufr/BufrParser/Exports/Variables/DatetimeVariable.cpp:91]: Once this PR is merged, I will likely take on the task of switching entirely to the newer dateTime with epoch format and eliminate these strings.  So for the time being, can we just move past this one?

File: src/bufr/DataObject/ArrayDataObject.cpp
Comment (srherbener) [reply src/bufr/DataObject/ArrayDataObject.cpp:54]: Thanks for the explanation @gthompsnJCSDA. I think it sounds good - I was not aware of a BUFR standard for missing values. I have a feeling that not all BUFR files from all providers will follow that standard but hopefully those are few and far between. I think it makes sense to go with the changes you made and address any missing values issues (if they come up) in a subsequent PR.

File: src/bufr/BufrParser/Exports/Variables/DatetimeVariable.cpp
Comment (srherbener) [reply src/bufr/BufrParser/Exports/Variables/DatetimeVariable.cpp:91]: Sure! I think that is a good pragmatic approach. I'm fine with deferring this to a subsequent PR.

**Review by srherbener — APPROVED** [review]

**Comment by srherbener** [comment]
@gthompsnJCSDA, @rmclaren is it okay to merge this after CI testing runs (and passes of course). The reason I'm asking is that @gthompsnJCSDA mentioned that these changes sync up with a feature branch (instead of develop branch) for the BUFR converter code. As long as you are okay with managing that situation, I'm okay with merging.

Thanks!

**Pull request merged**
